File: 2023/day3.py
from collections import defaultdict

with open("inputs/day3.txt") as file:
	puzzle = file.read().strip()

lines = puzzle.split("\n")
not_symbols = "0123456789."
digits = "0123456789"
result = 0
positions = []
for i in range(len(lines)):
	number = ""
	for j in range(len(lines[i])):
		if lines[i][j] not in digits and number != "":
			positions.append((i, int(number), j-len(number), j))
			number = ""
			continue

		if lines[i][j].isdigit():
			number += lines[i][j]

	if number:
		positions.append((i,int(number), j-len(number)+1, j))

# Part one
for p in positions:
	i, n, j, k = p
	if j > 0 and lines[i][j-1] not in not_symbols:
		result += n
		continue
	if lines[i][k] not in not_symbols:
		result += n
		continue
	for ii in range(max(j-1,0), k+1):	
		if i > 0 and lines[i-1][ii] not in not_symbols:
			break
		if i < len(lines)-1 and lines[i+1][ii] not in not_symbols: 
			break
	else:
		continue
	result += n
	
print(result)


# Part Two
symbol_positions = defaultdict(list)

for p in positions:
	i, n, j, k = p
	if j > 0 and lines[i][j-1] not in not_symbols:
		symbol_positions[(i,j-1)].append(n)
	if lines[i][k] not in not_symbols:
		symbol_positions[(i,k)].append(n)
		result += n
	for ii in range(max(j-1,0), k+1):	
		if i > 0 and lines[i-1][ii] not in not_symbols:
			symbol_positions[(i-1,ii)].append(n)
		if i < len(lines)-1 and lines[i+1][ii] not in not_symbols: 
			symbol_positions[(i+1,ii)].append(n)

result = 0
for s, nums in symbol_positions.items():
	if len(nums) == 2:
		result += nums[0] * nums[1]

print(result)

File: 2023/day3.c
#include <stdio.h>
#include <string.h>

#define MAX 10000
#define ADJACENT(row1, row2) (row1+1 == row2)||(row1-1 == row2)||(row1 == row2)
#define SAME_RANGE(s1,s2,e1,e2) s1 >= s2-1 && e1 <= e2

struct Part {
	int row;
	int start;
	int end;
	int number;
	char type;
};

struct Symbol {
	int number;
	int row;
	int col;
};


int digit_number(int number) {
	int d = 1;
	while ((number/=10)>0) {
		d++;
	}
	return d;
}

int main(int argc, char const *argv[])
{
	FILE *fp;
	char line[MAX];
	struct Part numlocs[MAX];
	int result = 0;

	fp = fopen("inputs/day3.txt", "r");
	if (fp == NULL) {
		perror("Error in opening file day3.txt");
		return 1;
	}

	int number = 0;
	int i = 0;
	int k = 0;
	int j = 0;
	while ((fgets(line, MAX, fp)) != NULL) {
		for (j = 0; line[j] != '\0'; j++) {
			if ((line[j] < '0' || line[j] > '9') && line[j] != '.' && line[j] != '\n') {
				struct Part n =  {i, j, j, line[j], 's'};
				numlocs[k++] = n;
			}
			if (line[j] >= '0' && line[j] <= '9') {
				number = (line[j]-'0') + number * (number>0?10:1);
			}
			if ((line[j] < '0' || line[j] > '9') && number > 0) {
				struct Part n =  {i, j-digit_number(number), j, number, 'n'};
				numlocs[k++] = n;
				number = 0;
			}
		}
		i++;
	}
	if (number) {
		struct Part n =  {i-1, j-digit_number(number), j, number, 'n'};
		numlocs[k++] = n;
		number = 0;
	}
	fclose(fp);


	int symbol_numbers[150][150];
	int count_symbol_numbers[150][150];
	for (int i = 0; i < k; i++) {
		if (numlocs[i].type == 'n'){
			for (int j = 0; j < k; j++) {
				if ((numlocs[j].type == 's') &&
					(ADJACENT(numlocs[i].row, numlocs[j].row)) &&
					(SAME_RANGE(numlocs[j].start, numlocs[i].start, numlocs[j].end, numlocs[i].end))) {
					if (symbol_numbers[numlocs[j].row][numlocs[j].start] == 0) {
						symbol_numbers[numlocs[j].row][numlocs[j].start] = 1;
					}
					symbol_numbers[numlocs[j].row][numlocs[j].start] *= numlocs[i].number;
					count_symbol_numbers[numlocs[j].row][numlocs[j].start] += 1;
					result += numlocs[i].number;
					break;
				}
			}
		}
	}
	printf("Part one: %d\n", result);
	result = 0;
	for (int i=0; i < 150;i++){
		for (int j =0; j < 150;j++) {
			if (count_symbol_numbers[i][j] == 2) {
				result += symbol_numbers[i][j];
			}
		}
	}
	printf("Part two: %d\n", result);
	return 0;
}